### ISSUE TITLE
internal string upgrade latin-1 => utf8 after s/// results in illegal utf8

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8780]
Migrated from [rt.perl.org#41530](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=41530) (status was 'resolved')

Searchable as RT41530$

**Comment by p5pRT** [comment]
### From detlev@xs4all.nl

#### Created by detlev@xs4all.nl



\# The following two lines result in an illegal string $str

&nbsp;  my $str = "a\\x\{a0\}a";        \# create a str with two a's and char in the range \\x80\.\.\.\\xFF
&nbsp;  $str =~ s/a/\\x\{100\}/g;       \# replace the a's with a unicode char \( >= \\x\{100\}\)

\# The internal problem in $str shows up in the following operations&#8203;:

&nbsp;  my $n = ord\( substr\($str\,1\,1\) \);    \# crash 1 \(under 5\.8\.8 on Windows as well as 5\.8\.6 on FreeBSD\)
&nbsp;  $str =~ s/\[\\t\\f\]/\_/g;               \# crash 2 \(only under 5\.8\.8 on Windows\)

crash 1 is&#8203;:
Malformed UTF\-8 character \(unexpected continuation byte 0x80\, with no preceding
start byte\) in ord at subst\_bug\.pl line 4\.

crash 2&#8203;:
Malformed UTF\-8 character \(fatal\) at subst\_bug\.pl line 5\.

Inspection of the intermediate stages of $str shows that the following happens&#8203;: the string needs to
be converted to utf8 when the two a's are replaced with the chars \\x\{100\}\. The character \\x\{a0\} also
needs to be converted to its utf8\-representation C2 A0\, but gets overlooked\.

The problem doesn't occur when the input string is "a\\x\{a0\}b" or "b\\x\{a0\}a" so that only one 
substitution
occurs\. In that case the \\x\{a0\} in the middle get properly upgraded\.

The details can be seen by running this code&#8203;:

&nbsp;    \#\!/usr/bin/perl \-w

&nbsp;    use strict;
&nbsp;    use warnings;
&nbsp;    use Encode qw\(is\_utf8\);

&nbsp;    my $str = "a\\x\{a0\}a";
&nbsp;    show\($str\);

&nbsp;    $str =~ s/a/\\x\{100\}/g;
&nbsp;    show\($str\);

&nbsp;    my $n = ord\( substr\($str\, 1\, 1\) \);
&nbsp;    print "n = $n\\n";

&nbsp;    $str =~ s/\[\\t\\f\]/\_/g;
&nbsp;    show\($str\);

&nbsp;    sub show \{
&nbsp;    my $var = shift;
&nbsp;    my $u8 = is\_utf8\($var\) ? "UTF\-8" : "latin1";

&nbsp;    my $len = length\($var\);

&nbsp;    my @&#8203;bytes = unpack\("C\*"\, $var\);
&nbsp;    print "len=$len\, enc=$u8\, C="\,
&nbsp;        join\(" "\, map \{ sprintf\("%02x"\, $\_\) \} @&#8203;bytes\)\, "\\n";
&nbsp;    \}

Output&#8203;:

&nbsp;    len=3\, enc=latin1\, C=61 a0 61
&nbsp;    len=3\, enc=UTF\-8\, C=c4 80 a0 c4 80
&nbsp;    Malformed UTF\-8 character \(unexpected continuation byte 0xa0\, with no preceding start byte\) in 
ord at \./utf8bug line 13\.
&nbsp;    n = 0
&nbsp;    Malformed UTF\-8 character \(fatal\) at \./utf8bug line 16\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.6:

Configured by rj at Thu Feb  3 20:42:58 CET 2005.

Summary of my perl5 (revision 5 version 8 subversion 6) configuration:
  Platform:
    osname=freebsd, osvers=4.10-release-p2, archname=i386-freebsd-64int
    uname='freebsd xs0.xs4all.nl 4.10-release-p2 freebsd 4.10-release-p2 #1: thu aug 12 12:43:13 cest 2004 kai@xs0.xs4all.nl:usrobjusrsrcsysgeneric i386 '
    config_args='-sde -Dprefix=/usr/local -Darchlib=/usr/local/lib/perl5/5.8.6/mach -Dprivlib=/usr/local/lib/perl5/5.8.6 -Dman3dir=/usr/local/lib/perl5/5.8.6/perl/man/man3 -Dman1dir=/usr/local/man/man1 -Dsitearch=/usr/local/lib/perl5/site_perl/5.8.6/mach -Dsitelib=/usr/local/lib/perl5/site_perl/5.8.6 -Dscriptdir=/usr/local/bin -Dsiteman3dir=/usr/local/lib/perl5/5.8.6/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Ui_malloc -Ui_iconv -Uinstallusrbinperl -Dcc=cc -Doptimize=-O -pipe  -Duseshrplib -Dccflags=-DAPPLLIB_EXP="/usr/local/lib/perl5/5.8.6/BSDPAN" -Ud_dosuid -Ui_gdbm -Dusethreads=n -Dusemymalloc=y -Duse64bitint'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=undef
    usemymalloc=y, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-DAPPLLIB_EXP="/usr/local/lib/perl5/5.8.6/BSDPAN" -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -pipe -I/usr/local/include',
    optimize='-O -pipe ',
    cppflags='-DAPPLLIB_EXP="/usr/local/lib/perl5/5.8.6/BSDPAN" -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='2.95.4 20020320 [FreeBSD]', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -Wl,-E -L/usr/local/lib'
    libpth=/usr/lib /usr/local/lib
    libs=-lgdbm -lm -lcrypt -lutil
    perllibs=-lm -lcrypt -lutil
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='  -Wl,-R/usr/local/lib/perl5/5.8.6/mach/CORE'
    cccdlflags='-DPIC -fPIC', lddlflags='-shared  -L/usr/local/lib'

Locally applied patches:
    SUIDPERLIO0 - fix PERLIO_DEBUG local root exploit (CAN-2005-0155)
    SUIDPERLIO1 - fix PERLIO_DEBUG buffer overflow (CAN-2005-0156)


@INC for perl v5.8.6:
    /usr/local/lib/perl5/site_perl/5.8.6/mach
    /usr/local/lib/perl5/site_perl/5.8.6
    /usr/local/lib/perl5/site_perl
    /usr/local/lib/perl5/5.8.6/BSDPAN
    /usr/local/lib/perl5/5.8.6/mach
    /usr/local/lib/perl5/5.8.6
    .


Environment for perl v5.8.6:
    HOME=/home/d/detlev
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/d/detlev/bin:/usr/local/bin:/bin:/usr/bin:/usr/games:.:/home/d/detlev/bat
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From ThB@gymel.com

#### Created by ThB@gymel.com



I can confirm this behaviour under perl 5\.10\.0\.

Also \- without modifying the original string&#8203;:

s/\.\.\./\.\.\./g;          \# triggers the bug

1 while s/\.\.\./\.\.\./;   \# works o\.k\.


A workaround would be to enforce internal utf8\-ness before substituting&#8203;:
Either by

utf8&#8203;::uprgrade\($str\);

or by

$str = substr\("\\x\{100\}"\.$str\, 1\);

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.10.0:

Configured by rurban at Mon Jun 30 16:03:19 GMT 2008.

Summary of my perl5 (revision 5 version 10 subversion 0 patch 34065) configuration:
  Platform:
    osname=cygwin, osvers=1.5.25(0.15642), archname=cygwin-thread-multi-64int
    uname='cygwin_nt-5.1 reini 1.5.25(0.15642) 2008-06-12 19:34 i686 cygwin '
    config_args='-de -Dmksymlinks -Dusethreads -Dmad=y -Dusedevel'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=y, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__ -fno-strict-aliasing -pipe -I/usr/local/include',
    optimize='-O3',
    cppflags='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__ -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='3.4.4 (cygming special, gdc 0.12, using dmd 0.125)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='g++', ldflags =' -Wl,--enable-auto-import -Wl,--export-all-symbols -Wl,--stack,8388608 -Wl,--enable-auto-image-base -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib /lib
    libs=-lgdbm -ldb -ldl -lcrypt -lgdbm_compat
    perllibs=-ldl -lcrypt
    libc=/usr/lib/libc.a, so=dll, useshrplib=true, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' --shared  -Wl,--enable-auto-import -Wl,--export-all-symbols -Wl,--stack,8388608 -Wl,--enable-auto-image-base -L/usr/local/lib'

Locally applied patches:
    MAINT34065
    CYG11 no-bs
    CYG12 no archlib in otherlibdirs
    CYG14 Dynaloader
    CYG15 static-Win32CORE
    Bug#55162 File::Spec::case_tolerant performance


@INC for perl 5.10.0:
    /usr/lib/perl5/5.10/i686-cygwin
    /usr/lib/perl5/5.10
    /usr/lib/perl5/site_perl/5.10/i686-cygwin
    /usr/lib/perl5/site_perl/5.10
    /usr/lib/perl5/vendor_perl/5.10/i686-cygwin
    /usr/lib/perl5/vendor_perl/5.10
    /usr/lib/perl5/vendor_perl/5.10
    /usr/lib/perl5/site_perl/5.8
    /usr/lib/perl5/vendor_perl/5.8
    .


Environment for perl 5.10.0:
    CYGWIN=ntsec smbntsec nobinmode
    HOME=/home/thomas
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/cygdrive/c/program files/imagemagick-6.3.7-q16:/cygdrive/c/bin:/cygdrive/c/usr:/cygdrive/c/usr/local/bin:/cygdrive/c/Windows/system32:/cygdrive/c/Windows:/cygdrive/c/Windows/System32/Wbem:/cygdrive/d/perl/bin:/usr/bin:/cygdrive/c/GnuPG:/cygdrive/c/Program Files/Compaq/Compaq Management Agents/Dmi/Win32/Bin:/cygdrive/c/Program Files/TortoiseSVN/bin:/cygdrive/t/texlive2007/bin/win32:/cygdrive/t/texlive2007/dviout:/cygdrive/c/gs/gs8.54/bin:/cygdrive/c/Program Files/CollabNet Subversion:/cygdrive/c/program files/bind9
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

I'm preparing a patch for this bug\, and I'm uncertain about the best way
to do it\.

First\, the bug is caused by the code not realizing that when you have
two strings that independently may be in utf8 or not\, that there are 4
cases to take care of\.  I mention this because the error of only taking
care of 3 of the cases occurs in other places in the code as well\.

The code does not consider the possibility that the replacement string
could be in utf8 when the source/target string isn't\.  Thus 

$latin1 =~ s/latin1/utf8/;

fails\.  The solution is to upgrade the variable to utf8\.  My dilemma is
whether to always do the upgrade when the replacement string is in utf8\,
or to do it only if the match succeeds\.  The difference can lead to
different results later\, as if there is no upgrade\, the scalar's
characters in the 128\-255 range will have different semantics than if
the upgrade takes place\.

I'm leaning towards doing the upgrade\, as I think we can infer from the
replacement string being in utf8 that the programmer intended that the
string have Unicode semantics\, even if it isn't in utf8\.  Therefore\,
it's better to do the upgrade to force those semantics\.

Is there a contrary opinion?

\-\-Karl Williamson

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, Apr 11\, 2010 at 09&#8203;:07&#8203;:08PM \-0700\, Karl Williamson via RT wrote&#8203;:
> I'm preparing a patch for this bug\, and I'm uncertain about the best way
> to do it\.
> 
> First\, the bug is caused by the code not realizing that when you have
> two strings that independently may be in utf8 or not\, that there are 4
> cases to take care of\.  I mention this because the error of only taking
> care of 3 of the cases occurs in other places in the code as well\.
> 
> The code does not consider the possibility that the replacement string
> could be in utf8 when the source/target string isn't\.  Thus 
> 
> $latin1 =~ s/latin1/utf8/;
> 
> fails\.  The solution is to upgrade the variable to utf8\.  My dilemma is
> whether to always do the upgrade when the replacement string is in utf8\,
> or to do it only if the match succeeds\.  The difference can lead to
> different results later\, as if there is no upgrade\, the scalar's
> characters in the 128\-255 range will have different semantics than if
> the upgrade takes place\.
> 
> I'm leaning towards doing the upgrade\, as I think we can infer from the
> replacement string being in utf8 that the programmer intended that the
> string have Unicode semantics\, even if it isn't in utf8\.  Therefore\,
> it's better to do the upgrade to force those semantics\.
> 
> Is there a contrary opinion?

Yes\!

I think it could could just a validly be argued that the programmer only
intended the utf8 upgrade for the cases that matched\. Which makes us even\.
Then I think the tie\-breaker is that we should try to be as conservative
as possible and only upgrade when we need to\.

\-\- 
In economics\, the exam questions are the same every year\.
They just change the answers\.

**Comment by p5pRT** [comment]
### From @khwilliamson

<details><summary>0001-PATCH-perl-41530-s-non-utf8-is-utf8-fails.patch</summary>

```diff
From 091d7e65c1fb5709b7b76e77c6eec3fa1bee7eb8 Mon Sep 17 00:00:00 2001
From: Karl Williamson <khw@khw-desktop.(none)>
Date: Tue, 11 May 2010 10:57:41 -0600
Subject: [PATCH] PATCH: [perl #41530] s/non-utf8/is-utf8/ fails.

When the replacement is in utf8, there was failure to upgrade the result
when the source and the pattern weren't in utf8.  This simply checks
that when there is a match that will lead to the replacement being done.
It then does the upgrade.  If this led to changes in the source, we redo
the match because pointers to saved buffers could have changed.  There
may be other cases where we don't need to redo the match, but I don't
know the code well-enough to easily figure it out.
---
 pp_hot.c     |   17 +++++++++++++++++
 t/re/subst.t |   23 ++++++++++++++++++++++-
 2 files changed, 39 insertions(+), 1 deletions(-)

diff --git a/pp_hot.c b/pp_hot.c
index ea949b8..ab36593 100644
--- a/pp_hot.c
+++ b/pp_hot.c
@@ -2126,6 +2126,7 @@ PP(pp_subst)
 	DIE(aTHX_ "%s", PL_no_modify);
     PUTBACK;
 
+  setup_match:
     s = SvPV_mutable(TARG, len);
     if (!SvPOKp(TARG) || SvTYPE(TARG) == SVt_PVGV)
 	force_on_match = 1;
@@ -2181,6 +2182,22 @@ PP(pp_subst)
 			 r_flags | REXEC_CHECKED);
     /* known replacement string? */
     if (dstr) {
+
+	/* Upgrade the source if the replacement is utf8 but the source is not,
+	 * but only if it matched; see
+	 * http://www.nntp.perl.org/group/perl.perl5.porters/2010/04/msg158809.html
+	 */
+	if (matched && DO_UTF8(dstr) && ! DO_UTF8(TARG)) {
+	    const STRLEN new_len = sv_utf8_upgrade(TARG);
+
+	    /* If the lengths are the same, the pattern contains only
+	     * invariants, can keep going; otherwise, various internal markers
+	     * could be off, so redo */
+	    if (new_len != len) {
+		goto setup_match;
+	    }
+	}
+
 	/* replacement needing upgrading? */
 	if (DO_UTF8(TARG) && !doutf8) {
 	     nsv = sv_newmortal();
diff --git a/t/re/subst.t b/t/re/subst.t
index 042f67a..bdf442e 100644
--- a/t/re/subst.t
+++ b/t/re/subst.t
@@ -7,7 +7,7 @@ BEGIN {
 }
 
 require './test.pl';
-plan( tests => 143 );
+plan( tests => 149 );
 
 $x = 'foo';
 $_ = "x";
@@ -614,3 +614,24 @@ fresh_perl_is( '$_="abcef"; s/bc|(.)\G(.)/$1 ? "[$1-$2]" : "XX"/ge; print' => 'a
   'bug: $var =~ s/$qr//e calling get-magic on $_ as well as $var',
  );
 }
+
+{ # Bug #41530; replacing non-utf8 with a utf8 causes problems
+    my $string = "a\x{a0}a";
+    my $sub_string = $string;
+    ok(! utf8::is_utf8($sub_string), "Verify that string isn't initially utf8");
+    $sub_string =~ s/a/\x{100}/g;
+    ok(utf8::is_utf8($sub_string),
+                        'Verify replace of non-utf8 with utf8 upgrades to utf8');
+    is($sub_string, "\x{100}\x{A0}\x{100}",
+                            'Verify #41530 fixed: replace of non-utf8 with utf8');
+
+    my $non_sub_string = $string;
+    ok(! utf8::is_utf8($non_sub_string),
+                                    "Verify that string isn't initially utf8");
+    $non_sub_string =~ s/b/\x{100}/g;
+    ok(! utf8::is_utf8($non_sub_string),
+            "Verify that failed substitute doesn't change string's utf8ness");
+    is($non_sub_string, $string,
+                        "Verify that failed substitute doesn't change string");
+}
+    
-- 
1.5.6.3


```
</details>

**Comment by p5pRT** [comment]
### From @rgs

Thanks\, applied to bleadperl as 3e462cdc2087ddf90984010fabd80c30db92bfa0

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon May 17 00&#8203;:56&#8203;:02 2010\, rafael wrote&#8203;:
> Thanks\, applied to bleadperl as 3e462cdc2087ddf90984010fabd80c30db92bfa0

For reference\, this was fixed up subsequently by commits c95ca9b8cd1 and
5c1648b\.

\-\- 

Father Chrysostomos
